clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/deploy/actions/AttachTitusServiceLoadBalancers.java

```diff
@@ -0,0 +1,107 @@
+
+import com.netflix.spinnaker.clouddriver.aws.deploy.ops.loadbalancer.TargetGroupLookupHelper;
+import com.netflix.spinnaker.clouddriver.saga.SagaCommand;
+import com.netflix.spinnaker.clouddriver.saga.flow.SagaAction;
+import com.netflix.spinnaker.clouddriver.saga.models.Saga;
+import com.netflix.spinnaker.clouddriver.titus.JobType;
+import com.netflix.spinnaker.clouddriver.titus.TitusClientProvider;
+import com.netflix.spinnaker.clouddriver.titus.client.TitusLoadBalancerClient;
+import com.netflix.spinnaker.clouddriver.titus.deploy.description.TitusDeployDescription;
+import com.netflix.spinnaker.clouddriver.titus.deploy.events.TitusLoadBalancerAttached;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AttachTitusServiceLoadBalancers
+    implements SagaAction<AttachTitusServiceLoadBalancers.AttachTitusServiceLoadBalancersCommand> {
+
+  private final TitusClientProvider titusClientProvider;
+
+  @Autowired
+  public AttachTitusServiceLoadBalancers(TitusClientProvider titusClientProvider) {
+    this.titusClientProvider = titusClientProvider;
+  }
+
+  @NotNull
+  @Override
+  public Result apply(@NotNull AttachTitusServiceLoadBalancersCommand command, @NotNull Saga saga) {
+    final TitusDeployDescription description = command.description;
+
+    if (JobType.isEqual(description.getJobType(), JobType.BATCH)) {
+      // Batch jobs don't get load balancers
+      return new Result();
+    }
+
+    TitusLoadBalancerClient loadBalancerClient =
+        titusClientProvider.getTitusLoadBalancerClient(
+            description.getCredentials(), description.getRegion());
+    if (loadBalancerClient == null) {
+      // TODO(rz): This definitely doesn't seem like something to casually skip?
+      saga.log("Unable to create load balancing client in target account/region");
+      return new Result();
+    }
+
+    TargetGroupLookupHelper.TargetGroupLookupResult targetGroups =
+        command.getTargetGroupLookupResult();
+
+    if (targetGroups != null) {
+      String jobUri = command.getJobUri();
+
+      targetGroups
+          .getTargetGroupARNs()
+          .forEach(
+              targetGroupArn -> {
+                loadBalancerClient.addLoadBalancer(jobUri, targetGroupArn);
+                saga.log("Attached %s to %s", targetGroupArn, jobUri);
+                saga.addEvent(
+                    new TitusLoadBalancerAttached(
+                        saga.getName(), saga.getId(), jobUri, targetGroupArn));
+              });
+
+      saga.log("Load balancers applied");
+    }
+
+    return new Result();
+  }
+
+  @EqualsAndHashCode(callSuper = true)
+  @Getter
+  public static class AttachTitusServiceLoadBalancersCommand extends SagaCommand {
```

Comment:
Why is this a static class inside this other class? Trying to wrap my brain around this structure.

Reply:
This is just an organizational thing.

If you need to modify anything for attaching load balancers to a Titus service, it's all in one place. Furthermore, the likelihood that two `SagaAction` implementations using the same `SagaCommand` is pretty unlikely, but if that were the case, it would then make sense to break that `SagaCommand` to its own file. Since these commands are only being used within a single `SagaAction`, it makes sense to me to embed it as an inner class.